---
title: "Customize The Plot"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction
An important feature of `ggcoverage` is the support of customization. This vignette will show how to customize the plot from the following two aspects:

* customize the theme
* add additional layer

<hr />

## Customize the theme
### Inner object
Before we start, let me introduce the internal structure of layers provided by `ggcoverage`. `ggcoverage` provides twelve layers, and `geom_coverage`, `geom_protein` are based solely on `ggplot2`, while the others are based on `ggplot2` and [patchwork](https://patchwork.data-imaginist.com/).

| Modules         | Type       | Description                                          | Object             |
|-----------------|------------|------------------------------------------------------|--------------------|
| geom_coverage   | Coverage   | Create genome coverage plot                          | ggplot2            |
| geom_base       | Annotation | Add base, base frequency and amino acid annotations  | ggplot2, patchwork |
| geom_cnv        | Annotation | Add CNV annotation                                   | ggplot2, patchwork |
| geom_gc         | Annotation | Add GC content annotation                            | ggplot2, patchwork |
| geom_gene       | Annotation | Add gene annotation                                  | ggplot2, patchwork |
| geom_transcript | Annotation | Add geneâ€™s transcripts annotation                    | ggplot2, patchwork |
| geom_peak       | Annotation | Add peak annotation                                  | ggplot2, patchwork |
| geom_ideogram   | Annotation | Add chromosome ideogram annotation                   | ggplot2, patchwork |
| geom_tad        | Annotation | Add contact map annotation                           | ggplot2, patchwork |
| geom_link       | Annotation | Add link annotation                                  | ggplot2, patchwork |
| geom_protein    | Coverage   | Create protein coverage plot                         | ggplot2            |
| geom_feature    | Annotation | Add feature annotation for genome/protein coverage   | ggplot2, patchwork |

By the way, all inner themes used are available in [theme_ggcoverage.R](https://github.com/showteeth/ggcoverage/blob/main/R/theme_ggcoverage.R)

<hr />

### Library
```{r library, message=FALSE, warning=FALSE}
# library
library(ggplot2)
library(patchwork)
library(tidyverse)
library(ggcoverage)
library(rtracklayer)
```

<hr />

### Prepare data
```{r prepare_data, warning=FALSE, fig.height = 6, fig.width = 8, fig.align = "center"}
# prepare gtf
gtf.file = system.file("extdata", "used_hg19.gtf", package = "ggcoverage")
gtf.gr = rtracklayer::import.gff(con = gtf.file, format = 'gtf')

# sample metadata
sample.meta = data.frame(SampleName=c('Chr18_MCF7_ER_1','Chr18_MCF7_ER_2','Chr18_MCF7_ER_3','Chr18_MCF7_input'),
                         Type = c("MCF7_ER_1","MCF7_ER_2","MCF7_ER_3","MCF7_input"),
                         Group = c("IP", "IP", "IP", "Input"))
sample.meta
# track folder
track.folder = system.file("extdata", "ChIP-seq", package = "ggcoverage")
# load bigwig file
track.df = LoadTrackFile(track.folder = track.folder, format = "bw", region = "chr18:76822285-76900000",
                         meta.info = sample.meta)
# check data
head(track.df)
track.df = track.df %>% dplyr::filter(Type %in% c("MCF7_ER_1", "MCF7_input"))
# create mark region
mark.region=data.frame(start=c(76822533),
                       end=c(76823743),
                       label=c("Promoter"))
# check data
mark.region
# create basic coverage plot
basic.coverage = ggcoverage(data = track.df, color = "auto", range.position = "out",
                            mark.region=mark.region, show.mark.label = FALSE)
basic.coverage
```

<hr />

### Customize ggplot2 object
Customize the plot generated by `ggcoverage`:
```{r customize_ggplot2, warning=FALSE, fig.height = 6, fig.width = 8, fig.align = "center"}
basic.coverage + 
  # add title
  labs(title = "chr18:76,822,285-76,900,000") +
  theme(plot.title=element_text(hjust=0.5)) + 
  # change color
  scale_fill_manual(values = c("MCF7_ER_1"="green", "MCF7_input"="yellow")) + 
  # add rect
  geom_rect(
      data = data.frame(start = 76840533, end = 76842533),
      aes_string(xmin = "start", xmax = "end", ymin = "0", ymax = "Inf"),
      fill = "red", alpha = 0.6
    )
```

<hr />

### Customize patchwork object
[patchwork](https://patchwork.data-imaginist.com/) is a ggplot2 extension, which is used to combine separate ggplots into the same graphic.

The example plot:
```{r customize_patchwork_example, warning=FALSE,fig.height = 7, fig.width = 8, fig.align = "center"}
# get consensus peak file
peak.file = system.file("extdata", "ChIP-seq", "consensus.peak", package = "ggcoverage")
# example plot
chip.coverage = basic.coverage + labs(title = "chr18:76,822,285-76,900,000") +
  theme(plot.title=element_text(hjust=0.5)) +
  geom_gene(gtf.gr=gtf.gr, arrow.length = 0.04,arrow.size=0.25,
            gene.size = 0.75,
            utr.size = 1.5,
            exon.size = 2.5,label.size = 2.5, plot.height = 0.3) +
  geom_peak(bed.file = peak.file, plot.height = 0.1) +
  geom_ideogram(genome = "hg19",plot.space = 0, plot.height = 0.15)
# get the class of chip.coverage
class(chip.coverage)
# output the plot
chip.coverage
```

The above plot is a nested combination of ggplot2 object. We can obtain the elements (**Pay attention to changes in object**):
```{r customize_patchwork_obtain, warning=FALSE,fig.height = 6, fig.width = 8, fig.align = "center"}
# obtain the track other than ideogram
chip.coverage[[1]]
# class
class(chip.coverage[[1]])
```

```{r customize_patchwork_obtain2, warning=FALSE,fig.height = 6, fig.width = 8, fig.align = "center"}
# obtain the track other than peak
chip.coverage[[1]][[1]]
# class
class(chip.coverage[[1]][[1]])
```

```{r customize_patchwork_obtain3, warning=FALSE,fig.height = 2, fig.width = 8, fig.align = "center"}
# obtain the gene track
chip.coverage[[1]][[1]][[2]]
# class
class(chip.coverage[[1]][[1]][[2]])
```

```{r customize_patchwork_obtain4, warning=FALSE,fig.height = 2, fig.width = 8, fig.align = "center"}
# obtain the peak track
chip.coverage[[1]][[2]]
# class
class(chip.coverage[[1]][[2]])
```


**Add another peak info**:
```{r customize_patchwork_add_peak, warning=FALSE,fig.height = 2, fig.width = 8, fig.align = "center"}
chip.coverage[[1]][[2]] + 
  # the size is the height of segment, controlled by peak.size in geom_peak
  geom_segment(
      data = data.frame(start = 76840533, end = 76842533),
      aes_string(x = "start", xend = "end", y = "1", yend = "1"),
      color = "red", size = 5
    )
```

<hr />

## Add additional layer
### Obtain the data
To add additional layer, we need to obtain the raw coverage data (for consistency). `ggcoverage` provides `GetPlotData` to obtain the data used to plot.
```{r get_raw_data}
# get coverage data, the layer number is four
coverage.data = GetPlotData(plot = chip.coverage, layer.num = 4)
# inspect data
head(coverage.data)
str(coverage.data)
```

<hr />

### Add layer
Here, I will create a new peak layer as an example (this is a sample example that does not depend on the raw data, but when you use your own data, you should use the raw data as region constraint).

```{r add_peak_data, warning=FALSE,fig.height = 2, fig.width = 8, fig.align = "center"}
# create pseudo-peak, you can load your peak file instead (be aware of 0-based/1-based)
pseudo_peak = data.frame(chr="chr18", start = 76840533, end = 76842533)
# get region constraint
plot.region.start <- coverage.data[1, "start"]
plot.region.end <- coverage.data[nrow(coverage.data), "end"]

# create plot
peak.plot = ggplot() + 
  geom_segment(
      data = pseudo_peak,
      aes_string(x = "start", xend = "end", y = "1", yend = "1"),
      color = "red", size = 5
    ) + 
  labs(y="Peak") + 
  theme_classic() + 
  theme(
      axis.line.y = element_blank(),
      axis.text.y = element_blank(),
      axis.title.y.right = element_text(color = "black", angle = 90, vjust = 0.5),
      axis.ticks.y = element_blank(),
      axis.text.x = element_blank(),
      axis.title.x = element_blank(),
      axis.ticks.x = element_blank(),
      panel.border = element_rect(colour = "black", fill = NA, size = 1),
      plot.margin = margin(t = 0.1, b = 0.1)
    ) + 
    scale_y_continuous(
      limits = c(1 - 0.1, 1 + 0.1),
      expand = c(0, 0), position = "right"
    ) + 
    scale_x_continuous(expand = c(0, 0)) + 
    coord_cartesian(xlim = c(plot.region.start, plot.region.end))
peak.plot
```


Combine the plot:
```{r combine_layers, warning=FALSE,fig.height = 7, fig.width = 8, fig.align = "center"}
# add peak layer
add.peak = patchwork::wrap_plots(chip.coverage[[1]] + theme(plot.margin = margin(t = 0.1, b = 0.1)),
    peak.plot,
    ncol = 1, heights = c(1, 0.1)
  )
add.peak
# add ideogram layers
final.plot = patchwork::wrap_plots(add.peak + theme(plot.margin = margin(t = 0.1, b = 0.1)),
    chip.coverage[[2]],
    ncol = 1, heights = c(1, 0.1)
  )
final.plot
```

<hr />


## Reference
* [ggplot2](https://ggplot2.tidyverse.org/)
* [patchwork](https://patchwork.data-imaginist.com/)

<br/>














