<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Customize The Plot • ggcoverage</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Customize The Plot">
<meta property="og:description" content="ggcoverage">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ggcoverage</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ggcoverage.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/CustomizeThePlot.html">Customize The Plot</a>
    </li>
    <li>
      <a href="../articles/TimeAndMemory.html">Time And Memory</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="CustomizeThePlot_files/header-attrs-2.14/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Customize The Plot</h1>
            
      
      
      <div class="hidden name"><code>CustomizeThePlot.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>An important feature of <code>ggcoverage</code> is the support of customization. This vignette will show how to customize the plot from the following two aspects:</p>
<ul>
<li>customize the elements of existing plot</li>
<li>add additional layer to the existing plot</li>
</ul>
<hr>
</div>
<div id="customize-the-theme" class="section level2">
<h2 class="hasAnchor">
<a href="#customize-the-theme" class="anchor"></a>Customize the theme</h2>
<div id="inner-object" class="section level3">
<h3 class="hasAnchor">
<a href="#inner-object" class="anchor"></a>Inner object</h3>
<p>Before we start, let me introduce the internal structure of layers provided by <code>ggcoverage</code>. <code>ggcoverage</code> provides twelve layers, and <code>geom_coverage</code>, <code>geom_protein</code> are based solely on <code>ggplot2</code>, while the others are based on <code>ggplot2</code> and <a href="https://patchwork.data-imaginist.com/">patchwork</a>.</p>
<table class="table">
<colgroup>
<col width="16%">
<col width="11%">
<col width="52%">
<col width="19%">
</colgroup>
<thead><tr class="header">
<th>Modules</th>
<th>Type</th>
<th>Description</th>
<th>Object</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>geom_coverage</td>
<td>Coverage</td>
<td>Create genome coverage plot</td>
<td>ggplot2</td>
</tr>
<tr class="even">
<td>geom_base</td>
<td>Annotation</td>
<td>Add base, base frequency and amino acid annotations</td>
<td>ggplot2, patchwork</td>
</tr>
<tr class="odd">
<td>geom_cnv</td>
<td>Annotation</td>
<td>Add CNV annotation</td>
<td>ggplot2, patchwork</td>
</tr>
<tr class="even">
<td>geom_gc</td>
<td>Annotation</td>
<td>Add GC content annotation</td>
<td>ggplot2, patchwork</td>
</tr>
<tr class="odd">
<td>geom_gene</td>
<td>Annotation</td>
<td>Add gene annotation</td>
<td>ggplot2, patchwork</td>
</tr>
<tr class="even">
<td>geom_transcript</td>
<td>Annotation</td>
<td>Add gene’s transcripts annotation</td>
<td>ggplot2, patchwork</td>
</tr>
<tr class="odd">
<td>geom_peak</td>
<td>Annotation</td>
<td>Add peak annotation</td>
<td>ggplot2, patchwork</td>
</tr>
<tr class="even">
<td>geom_ideogram</td>
<td>Annotation</td>
<td>Add chromosome ideogram annotation</td>
<td>ggplot2, patchwork</td>
</tr>
<tr class="odd">
<td>geom_tad</td>
<td>Annotation</td>
<td>Add contact map annotation</td>
<td>ggplot2, patchwork</td>
</tr>
<tr class="even">
<td>geom_link</td>
<td>Annotation</td>
<td>Add link annotation</td>
<td>ggplot2, patchwork</td>
</tr>
<tr class="odd">
<td>geom_protein</td>
<td>Coverage</td>
<td>Create protein coverage plot</td>
<td>ggplot2</td>
</tr>
<tr class="even">
<td>geom_feature</td>
<td>Annotation</td>
<td>Add feature annotation for genome/protein coverage</td>
<td>ggplot2, patchwork</td>
</tr>
</tbody>
</table>
<p>By the way, all inner themes used are available in <a href="https://github.com/showteeth/ggcoverage/blob/main/R/theme_ggcoverage.R">theme_ggcoverage.R</a></p>
<hr>
</div>
<div id="library" class="section level3">
<h3 class="hasAnchor">
<a href="#library" class="anchor"></a>Library</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># library</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ggcoverage</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></code></pre></div>
<hr>
</div>
<div id="prepare-data" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-data" class="anchor"></a>Prepare data</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># prepare gtf</span>
<span class="va">gtf.file</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"used_hg19.gtf"</span>, package <span class="op">=</span> <span class="st">"ggcoverage"</span><span class="op">)</span>
<span class="va">gtf.gr</span> <span class="op">=</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html">import.gff</a></span><span class="op">(</span>con <span class="op">=</span> <span class="va">gtf.file</span>, format <span class="op">=</span> <span class="st">'gtf'</span><span class="op">)</span>

<span class="co"># sample metadata</span>
<span class="va">sample.meta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>SampleName<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Chr18_MCF7_ER_1'</span>,<span class="st">'Chr18_MCF7_ER_2'</span>,<span class="st">'Chr18_MCF7_ER_3'</span>,<span class="st">'Chr18_MCF7_input'</span><span class="op">)</span>,
                         Type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MCF7_ER_1"</span>,<span class="st">"MCF7_ER_2"</span>,<span class="st">"MCF7_ER_3"</span>,<span class="st">"MCF7_input"</span><span class="op">)</span>,
                         Group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"IP"</span>, <span class="st">"IP"</span>, <span class="st">"IP"</span>, <span class="st">"Input"</span><span class="op">)</span><span class="op">)</span>
<span class="va">sample.meta</span></code></pre></div>
<pre><code>##         SampleName       Type Group
## 1  Chr18_MCF7_ER_1  MCF7_ER_1    IP
## 2  Chr18_MCF7_ER_2  MCF7_ER_2    IP
## 3  Chr18_MCF7_ER_3  MCF7_ER_3    IP
## 4 Chr18_MCF7_input MCF7_input Input</code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># track folder</span>
<span class="va">track.folder</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"ChIP-seq"</span>, package <span class="op">=</span> <span class="st">"ggcoverage"</span><span class="op">)</span>
<span class="co"># load bigwig file</span>
<span class="va">track.df</span> <span class="op">=</span> <span class="fu"><a href="../reference/LoadTrackFile.html">LoadTrackFile</a></span><span class="op">(</span>track.folder <span class="op">=</span> <span class="va">track.folder</span>, format <span class="op">=</span> <span class="st">"bw"</span>, region <span class="op">=</span> <span class="st">"chr18:76822285-76900000"</span>,
                         meta.info <span class="op">=</span> <span class="va">sample.meta</span><span class="op">)</span>
<span class="co"># check data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">track.df</span><span class="op">)</span></code></pre></div>
<pre><code>##   seqnames    start      end   score      Type Group
## 1    chr18 76820285 76820400 219.658 MCF7_ER_1    IP
## 2    chr18 76820401 76820700   0.000 MCF7_ER_1    IP
## 3    chr18 76820701 76821000 439.316 MCF7_ER_1    IP
## 4    chr18 76821001 76821300 219.658 MCF7_ER_1    IP
## 5    chr18 76821301 76821600   0.000 MCF7_ER_1    IP
## 6    chr18 76821601 76821900 219.658 MCF7_ER_1    IP</code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">track.df</span> <span class="op">=</span> <span class="va">track.df</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Type</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MCF7_ER_1"</span>, <span class="st">"MCF7_input"</span><span class="op">)</span><span class="op">)</span>
<span class="co"># create mark region</span>
<span class="va">mark.region</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>start<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">76822533</span><span class="op">)</span>,
                       end<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">76823743</span><span class="op">)</span>,
                       label<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Promoter"</span><span class="op">)</span><span class="op">)</span>
<span class="co"># check data</span>
<span class="va">mark.region</span></code></pre></div>
<pre><code>##      start      end    label
## 1 76822533 76823743 Promoter</code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create basic coverage plot</span>
<span class="va">basic.coverage</span> <span class="op">=</span> <span class="fu"><a href="../reference/ggcoverage.html">ggcoverage</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">track.df</span>, color <span class="op">=</span> <span class="st">"auto"</span>, range.position <span class="op">=</span> <span class="st">"out"</span>,
                            mark.region<span class="op">=</span><span class="va">mark.region</span>, show.mark.label <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">basic.coverage</span></code></pre></div>
<p><img src="CustomizeThePlot_files/figure-html/prepare_data-1.png" width="768" style="display: block; margin: auto;"></p>
<hr>
</div>
<div id="customize-ggplot2-object" class="section level3">
<h3 class="hasAnchor">
<a href="#customize-ggplot2-object" class="anchor"></a>Customize ggplot2 object</h3>
<p>Customize the plot generated by <code>ggcoverage</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">basic.coverage</span> <span class="op">+</span> 
  <span class="co"># add title</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"chr18:76,822,285-76,900,000"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.title<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>hjust<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co"># change color</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MCF7_ER_1"</span><span class="op">=</span><span class="st">"green"</span>, <span class="st">"MCF7_input"</span><span class="op">=</span><span class="st">"yellow"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co"># add rect</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_rect</a></span><span class="op">(</span>
      data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fl">76840533</span>, end <span class="op">=</span> <span class="fl">76842533</span><span class="op">)</span>,
      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes_.html">aes_string</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="st">"start"</span>, xmax <span class="op">=</span> <span class="st">"end"</span>, ymin <span class="op">=</span> <span class="st">"0"</span>, ymax <span class="op">=</span> <span class="st">"Inf"</span><span class="op">)</span>,
      fill <span class="op">=</span> <span class="st">"red"</span>, alpha <span class="op">=</span> <span class="fl">0.6</span>
    <span class="op">)</span></code></pre></div>
<pre><code>## Scale for fill is already present.
## Adding another scale for fill, which will replace the existing scale.</code></pre>
<p><img src="CustomizeThePlot_files/figure-html/customize_ggplot2-1.png" width="768" style="display: block; margin: auto;"></p>
<hr>
</div>
<div id="customize-patchwork-object" class="section level3">
<h3 class="hasAnchor">
<a href="#customize-patchwork-object" class="anchor"></a>Customize patchwork object</h3>
<p><a href="https://patchwork.data-imaginist.com/">patchwork</a> is a ggplot2 extension, which is used to combine separate ggplots into the same graphic.</p>
<p>The example plot:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get consensus peak file</span>
<span class="va">peak.file</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"ChIP-seq"</span>, <span class="st">"consensus.peak"</span>, package <span class="op">=</span> <span class="st">"ggcoverage"</span><span class="op">)</span>
<span class="co"># example plot</span>
<span class="va">chip.coverage</span> <span class="op">=</span> <span class="va">basic.coverage</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"chr18:76,822,285-76,900,000"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.title<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>hjust<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/geom_gene.html">geom_gene</a></span><span class="op">(</span>gtf.gr<span class="op">=</span><span class="va">gtf.gr</span>, arrow.length <span class="op">=</span> <span class="fl">0.04</span>,arrow.size<span class="op">=</span><span class="fl">0.25</span>,
            gene.size <span class="op">=</span> <span class="fl">0.75</span>,
            utr.size <span class="op">=</span> <span class="fl">1.5</span>,
            exon.size <span class="op">=</span> <span class="fl">2.5</span>,label.size <span class="op">=</span> <span class="fl">2.5</span>, plot.height <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/geom_peak.html">geom_peak</a></span><span class="op">(</span>bed.file <span class="op">=</span> <span class="va">peak.file</span>, plot.height <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/geom_ideogram.html">geom_ideogram</a></span><span class="op">(</span>genome <span class="op">=</span> <span class="st">"hg19"</span>,plot.space <span class="op">=</span> <span class="fl">0</span>, plot.height <span class="op">=</span> <span class="fl">0.15</span><span class="op">)</span></code></pre></div>
<pre><code>## Loading ideogram...</code></pre>
<pre><code>## Loading ranges...</code></pre>
<pre><code>## Scale for x is already present.
## Adding another scale for x, which will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get the class of chip.coverage</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">chip.coverage</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "patchwork" "gg"        "ggplot"</code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># output the plot</span>
<span class="va">chip.coverage</span></code></pre></div>
<p><img src="CustomizeThePlot_files/figure-html/customize_patchwork_example-1.png" width="768" style="display: block; margin: auto;"></p>
<p>The above plot is a nested combination of ggplot2 object. We can obtain the elements (<strong>Pay attention to changes in object</strong>):</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># obtain the track other than ideogram</span>
<span class="va">chip.coverage</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="CustomizeThePlot_files/figure-html/customize_patchwork_obtain-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># class</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">chip.coverage</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "patchwork" "gg"        "ggplot"</code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># obtain the track other than peak</span>
<span class="va">chip.coverage</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="CustomizeThePlot_files/figure-html/customize_patchwork_obtain2-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># class</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">chip.coverage</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "patchwork" "gg"        "ggplot"</code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># obtain the gene track</span>
<span class="va">chip.coverage</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="CustomizeThePlot_files/figure-html/customize_patchwork_obtain3-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># class</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">chip.coverage</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "gg"     "ggplot"</code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># obtain the peak track</span>
<span class="va">chip.coverage</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="CustomizeThePlot_files/figure-html/customize_patchwork_obtain4-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># class</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">chip.coverage</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "gg"     "ggplot"</code></pre>
<p><strong>Add another peak info</strong>:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">chip.coverage</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> 
  <span class="co"># the size is the height of segment, controlled by peak.size in geom_peak</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span><span class="op">(</span>
      data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fl">76840533</span>, end <span class="op">=</span> <span class="fl">76842533</span><span class="op">)</span>,
      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes_.html">aes_string</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"start"</span>, xend <span class="op">=</span> <span class="st">"end"</span>, y <span class="op">=</span> <span class="st">"1"</span>, yend <span class="op">=</span> <span class="st">"1"</span><span class="op">)</span>,
      color <span class="op">=</span> <span class="st">"red"</span>, size <span class="op">=</span> <span class="fl">5</span>
    <span class="op">)</span></code></pre></div>
<p><img src="CustomizeThePlot_files/figure-html/customize_patchwork_add_peak-1.png" width="768" style="display: block; margin: auto;"></p>
<hr>
</div>
</div>
<div id="add-additional-layer" class="section level2">
<h2 class="hasAnchor">
<a href="#add-additional-layer" class="anchor"></a>Add additional layer</h2>
<div id="obtain-the-data" class="section level3">
<h3 class="hasAnchor">
<a href="#obtain-the-data" class="anchor"></a>Obtain the data</h3>
<p>To add additional layer, we need to obtain the raw coverage data (for consistency). <code>ggcoverage</code> provides <code>GetPlotData</code> to obtain the data used to plot.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get coverage data, the layer number is four</span>
<span class="va">coverage.data</span> <span class="op">=</span> <span class="fu"><a href="../reference/GetPlotData.html">GetPlotData</a></span><span class="op">(</span>plot <span class="op">=</span> <span class="va">chip.coverage</span>, layer.num <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>
<span class="co"># inspect data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">coverage.data</span><span class="op">)</span></code></pre></div>
<pre><code>##   seqnames    start      end   score      Type Group
## 1    chr18 76820285 76820400 219.658 MCF7_ER_1    IP
## 2    chr18 76820401 76820700   0.000 MCF7_ER_1    IP
## 3    chr18 76820701 76821000 439.316 MCF7_ER_1    IP
## 4    chr18 76821001 76821300 219.658 MCF7_ER_1    IP
## 5    chr18 76821301 76821600   0.000 MCF7_ER_1    IP
## 6    chr18 76821601 76821900 219.658 MCF7_ER_1    IP</code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">coverage.data</span><span class="op">)</span></code></pre></div>
<pre><code>## 'data.frame':    431 obs. of  6 variables:
##  $ seqnames: Factor w/ 1 level "chr18": 1 1 1 1 1 1 1 1 1 1 ...
##  $ start   : int  76820285 76820401 76820701 76821001 76821301 76821601 76821901 76822201 76822801 76823101 ...
##  $ end     : int  76820400 76820700 76821000 76821300 76821600 76821900 76822200 76822800 76823100 76823400 ...
##  $ score   : num  220 0 439 220 0 ...
##  $ Type    : Factor w/ 2 levels "MCF7_ER_1","MCF7_input": 1 1 1 1 1 1 1 1 1 1 ...
##  $ Group   : chr  "IP" "IP" "IP" "IP" ...</code></pre>
<hr>
</div>
<div id="add-layer" class="section level3">
<h3 class="hasAnchor">
<a href="#add-layer" class="anchor"></a>Add layer</h3>
<p>Here, I will create a new peak layer as an example (this is a sample example that does not depend on the raw data, but when you use your own data, you should use the raw data as region constraint).</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create pseudo-peak, you can load your peak file instead (be aware of 0-based/1-based)</span>
<span class="va">pseudo_peak</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>chr<span class="op">=</span><span class="st">"chr18"</span>, start <span class="op">=</span> <span class="fl">76840533</span>, end <span class="op">=</span> <span class="fl">76842533</span><span class="op">)</span>
<span class="co"># get region constraint</span>
<span class="va">plot.region.start</span> <span class="op">&lt;-</span> <span class="va">coverage.data</span><span class="op">[</span><span class="fl">1</span>, <span class="st">"start"</span><span class="op">]</span>
<span class="va">plot.region.end</span> <span class="op">&lt;-</span> <span class="va">coverage.data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">coverage.data</span><span class="op">)</span>, <span class="st">"end"</span><span class="op">]</span>

<span class="co"># create plot</span>
<span class="va">peak.plot</span> <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span><span class="op">(</span>
      data <span class="op">=</span> <span class="va">pseudo_peak</span>,
      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes_.html">aes_string</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"start"</span>, xend <span class="op">=</span> <span class="st">"end"</span>, y <span class="op">=</span> <span class="st">"1"</span>, yend <span class="op">=</span> <span class="st">"1"</span><span class="op">)</span>,
      color <span class="op">=</span> <span class="st">"red"</span>, size <span class="op">=</span> <span class="fl">5</span>
    <span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>y<span class="op">=</span><span class="st">"Peak"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/theme_peak.html">theme_peak</a></span><span class="op">(</span>margin.len <span class="op">=</span> <span class="fl">0.1</span>, x.range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">plot.region.start</span>, <span class="va">plot.region.end</span><span class="op">)</span><span class="op">)</span>
<span class="va">peak.plot</span></code></pre></div>
<p><img src="CustomizeThePlot_files/figure-html/add_peak_data-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Combine the plot:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># add peak layer</span>
<span class="va">add.peak</span> <span class="op">=</span> <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html">wrap_plots</a></span><span class="op">(</span><span class="va">chip.coverage</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">0.1</span>, b <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span>,
    <span class="va">peak.plot</span>,
    ncol <span class="op">=</span> <span class="fl">1</span>, heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.1</span><span class="op">)</span>
  <span class="op">)</span>
<span class="va">add.peak</span></code></pre></div>
<p><img src="CustomizeThePlot_files/figure-html/combine_layers-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># add ideogram layers</span>
<span class="va">final.plot</span> <span class="op">=</span> <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html">wrap_plots</a></span><span class="op">(</span><span class="va">add.peak</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">0.1</span>, b <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span>,
    <span class="va">chip.coverage</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,
    ncol <span class="op">=</span> <span class="fl">1</span>, heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.1</span><span class="op">)</span>
  <span class="op">)</span>
<span class="va">final.plot</span></code></pre></div>
<p><img src="CustomizeThePlot_files/figure-html/combine_layers-2.png" width="768" style="display: block; margin: auto;"></p>
<hr>
</div>
</div>
<div id="reference" class="section level2">
<h2 class="hasAnchor">
<a href="#reference" class="anchor"></a>Reference</h2>
<ul>
<li><a href="https://ggplot2.tidyverse.org/">ggplot2</a></li>
<li><a href="https://patchwork.data-imaginist.com/">patchwork</a></li>
</ul>
<p><br></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
