<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Time And Memory • ggcoverage</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Time And Memory">
<meta property="og:description" content="ggcoverage">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ggcoverage</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ggcoverage.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/CustomizeThePlot.html">Customize The Plot</a>
    </li>
    <li>
      <a href="../articles/TimeAndMemory.html">Time And Memory</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="TimeAndMemory_files/header-attrs-2.14/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Time And Memory</h1>
            
      
      
      <div class="hidden name"><code>TimeAndMemory.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>For large track files, it’s usually time and memory consuming to create coverage plot. <code>ggcoverage</code> provides two memory and time-efficient ways:</p>
<ul>
<li>For <strong>bigwig files</strong> and <strong>bam files that do not require normalization</strong>: <code>ggcoverage</code> loads the visualized region specified by users <strong>instead of</strong> loading the whole files and then extracting the visualized region.</li>
<li>For <strong>bam files that require normalization</strong>: <code>ggcoverage</code> utilizes <a href="https://bioconductor.org/packages/release/bioc/html/BiocParallel.html">BiocParallel</a> to <strong>perform normalization parallelly</strong>.</li>
</ul>
<hr>
</div>
<div id="load-region" class="section level2">
<h2 class="hasAnchor">
<a href="#load-region" class="anchor"></a>Load region</h2>
<div id="test-data" class="section level3">
<h3 class="hasAnchor">
<a href="#test-data" class="anchor"></a>Test data</h3>
<p>Here, we will load a big BAM file (~27G) from 10x single cell RNA-seq:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> possorted_genome_bam.bam</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -rw-r--r--. 1 songyabing wanglab 27G 8月  31 2021 possorted_genome_bam.bam</span></span></code></pre></div>
<hr>
</div>
<div id="load-region-1" class="section level3">
<h3 class="hasAnchor">
<a href="#load-region-1" class="anchor"></a>Load region</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ggcoverage</span><span class="op">)</span></code></pre></div>
<pre><code>## Registered S3 method overwritten by 'GGally':
##   method from   
##   +.gg   ggplot2</code></pre>
<pre><code>## Warning: replacing previous import 'ggplot2::annotate' by 'ggpp::annotate' when
## loading 'ggcoverage'</code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># prepare metadata</span>
<span class="va">sample.meta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>SampleName<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'possorted_genome_bam'</span><span class="op">)</span>,
                         Type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"possorted_genome_bam"</span><span class="op">)</span>,
                         Group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"10x"</span><span class="op">)</span><span class="op">)</span>
<span class="va">sample.meta</span></code></pre></div>
<pre><code>##             SampleName                 Type Group
## 1 possorted_genome_bam possorted_genome_bam   10x</code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># prepare track folder</span>
<span class="va">track.folder</span> <span class="op">=</span> <span class="st">'~/projects/ggcoverage'</span>
<span class="co"># load the track</span>
<span class="co"># region length: 3631</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span><span class="op">(</span><span class="va">track.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LoadTrackFile.html">LoadTrackFile</a></span><span class="op">(</span>
  track.folder <span class="op">=</span> <span class="va">track.folder</span>, format <span class="op">=</span> <span class="st">"bam"</span>, norm.method <span class="op">=</span> <span class="st">"None"</span>,
  region <span class="op">=</span> <span class="st">"chr11:118339075-118342705"</span>,
  extend <span class="op">=</span> <span class="fl">0</span>, meta.info <span class="op">=</span> <span class="va">sample.meta</span>
<span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## Calculate coverage with GenomicAlignments when norm.method is None!</code></pre>
<pre><code>##  用户  系统  流逝 
## 0.482 0.011 0.562</code></pre>
<p>check the data:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">track.df</span><span class="op">)</span></code></pre></div>
<pre><code>##   seqnames     start       end score                 Type Group
## 1    chr11 118339075 118339075   136 possorted_genome_bam   10x
## 2    chr11 118339076 118339076   169 possorted_genome_bam   10x
## 3    chr11 118339077 118339077   187 possorted_genome_bam   10x
## 4    chr11 118339078 118339078   233 possorted_genome_bam   10x
## 5    chr11 118339079 118339079   257 possorted_genome_bam   10x
## 6    chr11 118339080 118339080   281 possorted_genome_bam   10x</code></pre>
<p>For larger region:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># region length: 203631</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span><span class="op">(</span><span class="fu"><a href="../reference/LoadTrackFile.html">LoadTrackFile</a></span><span class="op">(</span>
  track.folder <span class="op">=</span> <span class="va">track.folder</span>, format <span class="op">=</span> <span class="st">"bam"</span>, norm.method <span class="op">=</span> <span class="st">"None"</span>,
  region <span class="op">=</span> <span class="st">"chr11:118339075-118542705"</span>,
  extend <span class="op">=</span> <span class="fl">0</span>, meta.info <span class="op">=</span> <span class="va">sample.meta</span>
<span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## Calculate coverage with GenomicAlignments when norm.method is None!</code></pre>
<pre><code>##  用户  系统  流逝 
## 0.862 0.032 0.900</code></pre>
<hr>
</div>
<div id="create-coverage" class="section level3">
<h3 class="hasAnchor">
<a href="#create-coverage" class="anchor"></a>Create coverage</h3>
<p>With the track dataframe, we can generate coverage plot.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create basic coverage plot</span>
<span class="co"># the running time is very small</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span><span class="op">(</span><span class="va">basic.coverage</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggcoverage.html">ggcoverage</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">track.df</span>, color <span class="op">=</span> <span class="st">"red"</span>, 
                                        range.position <span class="op">=</span> <span class="st">"out"</span>, show.mark.label <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##  用户  系统  流逝 
## 0.046 0.001 0.047</code></pre>
<p>The coverage plot (the bars are relatively dense, and can be viewed more clearly when saved as a PDF): <img src="../man/figures/time_memory_large.png" width="568" style="display: block; margin: auto;"></p>
<hr>
</div>
</div>
<div id="parallel-normalization" class="section level2">
<h2 class="hasAnchor">
<a href="#parallel-normalization" class="anchor"></a>Parallel normalization</h2>
<div id="test-data-1" class="section level3">
<h3 class="hasAnchor">
<a href="#test-data-1" class="anchor"></a>Test data</h3>
<p>To test the perfromance, we use three identical bam files from <a href="https://trace.ncbi.nlm.nih.gov/Traces/?run=SRR053616">SRR053616</a>.</p>
<p>Data folder for <strong>sequential normalization</strong>:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># test the sequential normalization</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> ./test</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; total 2.0G</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -rw-r--r--. 1 songyabing wanglab 3.9M May  26 16:44 SRR054616_rep3.bam.bai</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -rw-r--r--. 1 songyabing wanglab 3.9M May  26 16:44 SRR054616_rep2.bam.bai</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -rw-r--r--. 1 songyabing wanglab 3.9M May  26 16:44 SRR054616_rep1.bam.bai</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -rw-r--r--. 1 songyabing wanglab 646M May  26 16:44 SRR054616_rep3.bam</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -rw-r--r--. 1 songyabing wanglab 646M May  26 16:44 SRR054616_rep2.bam</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -rw-r--r--. 1 songyabing wanglab 646M May  26 16:44 SRR054616_rep1.bam</span></span></code></pre></div>
<p>Data folder for <strong>parallel normalization</strong>:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># test the parallel normalization</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> ./test2</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; total 2.0G</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -rw-r--r--. 1 songyabing wanglab 3.9M May  26 16:44 SRR054616_rep3.bam.bai</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -rw-r--r--. 1 songyabing wanglab 3.9M May  26 16:44 SRR054616_rep2.bam.bai</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -rw-r--r--. 1 songyabing wanglab 3.9M May  26 16:44 SRR054616_rep1.bam.bai</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -rw-r--r--. 1 songyabing wanglab 646M May  26 16:44 SRR054616_rep3.bam</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -rw-r--r--. 1 songyabing wanglab 646M May  26 16:44 SRR054616_rep2.bam</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -rw-r--r--. 1 songyabing wanglab 646M May  26 16:44 SRR054616_rep1.bam</span></span></code></pre></div>
<hr>
</div>
<div id="sequential-normalization" class="section level3">
<h3 class="hasAnchor">
<a href="#sequential-normalization" class="anchor"></a>Sequential normalization</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># prepare sample metadata</span>
<span class="va">sample.meta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>SampleName<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'SRR054616_rep1'</span>,<span class="st">'SRR054616_rep2'</span>,<span class="st">'SRR054616_rep3'</span><span class="op">)</span>,
                         Type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"SRR054616_rep1"</span>,<span class="st">"SRR054616_rep2"</span>,<span class="st">"SRR054616_rep3"</span><span class="op">)</span>,
                         Group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"rep1"</span>, <span class="st">"rep2"</span>, <span class="st">"rep3"</span><span class="op">)</span><span class="op">)</span>
<span class="va">sample.meta</span>

<span class="co"># track folder</span>
<span class="va">track.folder</span> <span class="op">=</span> <span class="st">"./test"</span>

<span class="co"># run</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span><span class="op">(</span><span class="va">track.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LoadTrackFile.html">LoadTrackFile</a></span><span class="op">(</span>
    track.folder <span class="op">=</span> <span class="va">track.folder</span>, format <span class="op">=</span> <span class="st">"bam"</span>, norm.method <span class="op">=</span> <span class="st">"RPKM"</span>,
    region <span class="op">=</span> <span class="st">"14:21,677,306-21,737,601"</span>, bamcoverage.path <span class="op">=</span> <span class="st">"~/anaconda3/bin/bamCoverage"</span>,
    extend <span class="op">=</span> <span class="fl">2000</span>, meta.info <span class="op">=</span> <span class="va">sample.meta</span>
  <span class="op">)</span><span class="op">)</span>

<span class="co">#&gt; user       system elapsed</span>
<span class="co">#&gt; 1169.767   35.005 1208.704</span></code></pre></div>
<p>The <strong>elapsed</strong> time (unit: seconds) is the time spent from the start to the end of the command.</p>
<hr>
</div>
<div id="parallel-normalization-1" class="section level3">
<h3 class="hasAnchor">
<a href="#parallel-normalization-1" class="anchor"></a>Parallel normalization</h3>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># prepare sample metadata</span>
<span class="va">sample.meta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>SampleName<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'SRR054616_rep1'</span>,<span class="st">'SRR054616_rep2'</span>,<span class="st">'SRR054616_rep3'</span><span class="op">)</span>,
                         Type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"SRR054616_rep1"</span>,<span class="st">"SRR054616_rep2"</span>,<span class="st">"SRR054616_rep3"</span><span class="op">)</span>,
                         Group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"rep1"</span>, <span class="st">"rep2"</span>, <span class="st">"rep3"</span><span class="op">)</span><span class="op">)</span>
<span class="va">sample.meta</span>

<span class="co"># track folder</span>
<span class="va">track.folder</span> <span class="op">=</span> <span class="st">"./test2"</span>

<span class="co"># run with three cores</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span><span class="op">(</span><span class="va">track.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LoadTrackFile.html">LoadTrackFile</a></span><span class="op">(</span>
  track.folder <span class="op">=</span> <span class="va">track.folder</span>, format <span class="op">=</span> <span class="st">"bam"</span>, norm.method <span class="op">=</span> <span class="st">"RPKM"</span>,
  region <span class="op">=</span> <span class="st">"14:21,677,306-21,737,601"</span>, bamcoverage.path <span class="op">=</span> <span class="st">"~/anaconda3/bin/bamCoverage"</span>,
  extend <span class="op">=</span> <span class="fl">2000</span>, meta.info <span class="op">=</span> <span class="va">sample.meta</span>, n.cores <span class="op">=</span> <span class="fl">3</span>
<span class="op">)</span><span class="op">)</span>

<span class="co">#&gt; user       system  elapsed</span>
<span class="co">#&gt; 1191.628   36.818  424.143</span></code></pre></div>
<p><mark>Here we can see, the time spent in parallel normalization is nearly one-third that of sequential normalization.</mark></p>
<hr>
</div>
<div id="create-coverage-1" class="section level3">
<h3 class="hasAnchor">
<a href="#create-coverage-1" class="anchor"></a>Create coverage</h3>
<p>In general, the loading step is the most time-consuming step. With the track dataframe, we can generate coverage plot.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># read the track dataframe</span>
<span class="va">track.df</span> <span class="op">=</span> <span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"~/projects/ggcoverage/ggcoverage_parallel_track.txt"</span>, 
                             sep <span class="op">=</span> <span class="st">"\t"</span>, header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
<span class="co"># create basic coverage plot</span>
<span class="co"># the running time is very small</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span><span class="op">(</span><span class="va">basic.coverage</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggcoverage.html">ggcoverage</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">track.df</span>, color <span class="op">=</span> <span class="st">"auto"</span>, 
                                        range.position <span class="op">=</span> <span class="st">"out"</span>, show.mark.label <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning in geom_coverage(data = data, mapping = mapping, color = color, : The
## color you provided is not as long as Type column in data, select automatically!</code></pre>
<pre><code>##  用户  系统  流逝 
## 0.018 0.000 0.018</code></pre>
<p>The coverage plot: <img src="TimeAndMemory_files/figure-html/parallel_coverage_plot-1.png" width="1152" style="display: block; margin: auto;"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
